include/nbl/video/IDescriptorPool.h

```diff
@@ -32,86 +30,186 @@ class NBL_API IDescriptorPool : public core::IReferenceCounted, public IBackendO
-            m_storageImageStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUImageView>>[]>(m_maxDescriptorCount[asset::EDT_STORAGE_IMAGE] + m_maxDescriptorCount[asset::EDT_INPUT_ATTACHMENT]);
-            m_UBO_SSBOStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBuffer>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER] + m_maxDescriptorCount[asset::EDT_UNIFORM_BUFFER_DYNAMIC] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_UTB_STBStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUBufferView>>[]>(m_maxDescriptorCount[asset::EDT_UNIFORM_TEXEL_BUFFER] + m_maxDescriptorCount[asset::EDT_STORAGE_BUFFER_DYNAMIC]);
-            m_accelerationStructureStorage = std::make_unique<core::StorageTrivializer<core::smart_refctd_ptr<IGPUAccelerationStructure>>[]>(m_maxDescriptorCount[asset::EDT_ACCELERATION_STRUCTURE]);
-        }
-
-        virtual ~IDescriptorPool() {}
-
-        virtual core::smart_refctd_ptr<IGPUDescriptorSet> createDescriptorSet_impl(core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& layout) = 0;
-        virtual bool freeDescriptorSets_impl(const uint32_t descriptorSetCount, IGPUDescriptorSet* const* const descriptorSets) = 0;
+            inline void reset(const bool allowsFreeing)
+            {
+                if (!allowsFreeing)
+                    linearAllocator.reset();
+                else
+                    generalAllocator.reset();
+            }
 
-        uint32_t m_maxSets;
+            inline uint32_t getAllocatedDescriptorCount(const bool allowsFreeing) const
+            {
+                if (!allowsFreeing)
+                    return linearAllocator.get_allocated_size();
+                else
+                    return generalAllocator.get_allocated_size();
+            }
 
-    private:
-        const IDescriptorPool::E_CREATE_FLAGS m_flags;
-        uint32_t m_maxDescriptorCount[asset::EDT_COUNT];
-        union
-        {
-            core::LinearAddressAllocator<uint32_t> m_linearAllocators[asset::EDT_COUNT+1];
-            core::GeneralpurposeAddressAllocator<uint32_t> m_generalAllocators[asset::EDT_COUNT+1];
+            union
+            {
+                core::LinearAddressAllocator<uint32_t> linearAllocator;
+                core::GeneralpurposeAddressAllocator<uint32_t> generalAllocator;
+            };
+            std::unique_ptr<uint8_t[]> generalAllocatorReservedSpace = nullptr;
         };
-        std::unique_ptr<uint8_t[]> m_generalAllocatorReservedSpace[asset::EDT_COUNT+1];
+        std::unique_ptr<allocator_state_t> m_descriptorAllocators[static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT) + 1];
+
+        const SCreateInfo m_creationParameters;
+        std::atomic_uint32_t m_version;
```

Comment:
lets add another atomic counter `m_currentlyAllocatedDescriptors` for implementing https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1115559849

Reply:
actually according to https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/8fbc87079edc5109a9c916eeb7ea7c0eff9ce128..b3719ddb1377d8af8f8d5f4dd112d3c844394553#r1116363532 that might be the only counter needed

Reply:
no need for atomic counter anymore, can derive from `IteratablePoolAddressAllocator`'s allocation count

Reply:
so basically the `m_version` doesn't need to exist, or any other atomic counter

Reply:
Resolved.